### DISCUSSION TITLE
Please share your use case of Rod

### DISCUSSION THREAD

**Discussion by ysmood** [go-rod/rod#412] [Q&A]
Please leave a comment here to share what you use Rod for. It will be great if you could share your project link.

Some popular projects that are using Rod:

- https://github.com/charmbracelet/vhs
- https://github.com/authelia/authelia
- https://github.com/projectdiscovery/nuclei
- https://github.com/projectdiscovery/katana
- https://github.com/semanser/codel
- https://github.com/philippta/flyscrape
- https://github.com/storj/storj
- https://github.com/kubeshop/botkube
- https://github.com/projectdiscovery/httpx
- https://github.com/Hackmanit/TInjA
- https://github.com/sigstore/sigstore
- https://github.com/SleepingBag945/dddd
- https://github.com/FrenchGithubUser/Hatt
- https://github.com/yaklang/yaklang
- https://github.com/janpfeifer/gonb
- https://github.com/sensepost/gowitness

**Comment by datoga** [comment]
For JS execution and data extraction, as puppeteer replacer, with better performance and cleaner API.

**Comment by imag** [comment]
Writing automation tests and data extraction

**Comment by BenjaminB64** [comment]
To extract cookie and localStorage from a website

**Reply by samueltuyizere** [reply]
@BenjaminB64  How did this go?

**Comment by gusanmaz** [comment]
Very recently created a CLI screenshot utility using Rod. 
You could check it out here: [Webshot](https://github.com/gusanmaz/webshot)

**Reply by dhidaa** [reply]
nice work man

**Reply by gusanmaz** [reply]
Thanks!

**Comment by mazihao66** [comment]
html to pdf

**Reply by Wiandono** [reply]
I second this

**Comment by dhidaa** [comment]
I used it to collect bank mutation data, thanks

**Comment by elbek** [comment]
I use it to collect available loads for booking in transportation area

**Comment by jaekook-neonesia** [comment]
To extract transaction history of indonesia web banking .

**Comment by Bolado** [comment]
Automation of long tasks in web applications

**Comment by audstanley** [comment]
Our team at CSUF is building a terminal UI application for windows, Mac, Linux, x86, arm64, arm64, and arm8 (ie: raspberry pi computers) to download books from LibGen without using the browser or API. It's a school project for a software testing course.  I might even see if we can build in the pi functionality into go-rod [since go-rod needs to connect to the chromium web socket via a separate/or/subprocess] than make a pull request... hopefully... if time will permit.  I think having platform/architecture detection during compile time is a good approach for go-rod's future (at least for a raspberry pi system, specifically). 🙂 We love this library, so far.  Some really cool things could be built on a pi will full-blown pi compatibility at compile time.  Would the go-rod team be interested in a dedicated "self-hosted" GitHub action runner on a pi4 (for builds/tests)? I haven't tested runners on a pi 4, but our team plans on it.

Edit:

I've just ran a dedicated GitHub action runner on a pi 4, and it's awesome (I've got a pi 0,2&3 - so I'll have to set those up also).  For now, i think we will work around the fact that the raspberry pi doesn't work with the chrome binaries that go-rod gets from the web, but I've forked the go-rod repo, so hopefully this can be an enhancement for go-rod at some point, if the team is interested in such an undertaking.

**Reply by audstanley** [reply]
...I took a stab at recursive html parsing with Google's x/net/html... And, oh boy, it's far too time consuming for the work ahead of us (amazing, but we just can't).

**Reply by ysmood** [reply]
Sounds interesting, how about open a formal issue to discuss this?

**Reply by audstanley** [reply]
will do!  I'm going to run some tests (on our project), but also... i really need to figure out how to run a subprocess in go, wait about 3 seconds, grab some text from both the stdin, and stderr pipes... regex out web-socket from the pipe [all while keeping the subprocess open during the duration of rod.New()]... then i think **for sure** I can get **chromium-browser** to launch as a subprocess, and have go-rod auto connect to the websocket [on the assumption go-rod is running on a pi].  For the most part, this should do the trick.  I don't know if this is too "hacky" of a solution, but it *should* work (if nothing else, than for our use case).  I'll open an issue soon, and try to make the details clear after also running some tests.  I've forked go-rod, so I might also be able to figure out where this subprocess can be run during the rod.New() constructor.  i would love to be able to contribute to this killer library. :slightly_smiling_face:

**Reply by ysmood** [reply]
I think rod already supports this kind of customization well, check this example file https://github.com/go-rod/rod/blob/master/lib/launcher/example_test.go

**Reply by audstanley** [reply]
No way... I'll test this out rn.

Edit:
[This worked perfectly for the pi 4 - and likely will work for other pies](https://github.com/go-rod/rod/blob/026d63501e51f0494ce1121972cf8be3321a958d/lib/launcher/example_test.go#L13)

**Comment by axetroy** [comment]
Here is a CLI tool for downloading video from the website and it uses rod for getting the cookie

https://github.com/iawia002/lux/blob/0c6b59388707bd386559261430c65db43da3d41e/cookier/cookier.go#L7-L22

**Comment by zakiego** [comment]
I use it to fill class attendance (requires authentication, so it can't only use http, must use a browser), and download data from [Tableau](https://github.com/zakiego/dataset-testing-kabupaten-kota-covid19/blob/main/script/download_data.go)

**Comment by danieldids** [comment]
Sustainability related Brazilian norms crawler (in progress).

I'm using go-colly and go-rod for this project. Temporary URL: https://normalis.mooo.com/

**Comment by shepherdjerred** [comment]
We're using go-rod with [testcontainers](https://golang.testcontainers.org) to automate end-to-end tests for our software.

**Comment by veggiemonk** [comment]
There is around 700 projects that uses `rod` 
https://github.com/search?q=path%3A**%2Fgo.mod+%22go-rod%2Frod%22&type=code

I mainly use it to login to website that don't have an API available.

**Comment by AndersonEngenheiro** [comment]
Im using to web scraping data games, movies and like bot to automatation a part of my real job.

**Comment by J-Siu** [comment]
I moved my pet project from selenium to rod.

An infinite scroll processor: https://github.com/J-Siu/go-is
A cli tool for scraping YT some page: https://github.com/J-Siu/yt-toolbox

**Comment by frontendr** [comment]
HTML / URL to PDF converter as a Docker container: https://github.com/frontendr/go-pdf

Awesome library!
